### PR TITLE
Array type implementation

### PR DESCRIPTION
**Pull request by sandreenko**
Array UDT.
Non-MD array has BaseClass, Count and ArrayElementType fields.
MD arrays also have each dimension length and lower bound values.

### CODE REVIEW

**Comment by sandreenko** [comment]
PTAL @nattress @AndyAyersMS

File: lib/ObjWriter/typeBuilder.cpp [hunk 1]
```diff
@@ -130,10 +132,91 @@ unsigned UserDefinedTypesBuilder::GetCompleteClassTypeIndex(
...
+  DataMemberRecord CountDMR(Access, IndexType, Offset, "count");
+  FLBR.writeMemberType(CountDMR);
+  FieldsCount++;
+  Offset += TargetPointerSize;
+
+  if (ArrayDescriptor.IsMultiDimensional == 1) {
+    for (unsigned i = 0; i < ArrayDescriptor.Rank; ++i) {
+      std::stringstream ss;
+      ss << "length" << i;
+      char *LengthName = new char[ss.gcount() + 1];
```
Comment (nattress) [inline lib/ObjWriter/typeBuilder.cpp]: This looks like a memory leak. You can probably just use ss.str().c_str() here to pass a char * through to the debug record serializer. Most string passing is done in LLVM with `llvm::StringRef` which might be more appropriate here, too.

File: lib/ObjWriter/typeBuilder.cpp [hunk 2]
```diff
@@ -130,10 +132,91 @@ unsigned UserDefinedTypesBuilder::GetCompleteClassTypeIndex(
...
+                                 Offset, LengthName);
+      FLBR.writeMemberType(LengthDMR);
+      FieldsCount++;
+      Offset += 4;
+    }
+
+    for (unsigned i = 0; i < ArrayDescriptor.Rank; ++i) {
+      std::stringstream ss;
+      ss << "bounds" << i;
+      char *BoundsName = new char[ss.gcount() + 1];
```
Comment (nattress) [inline lib/ObjWriter/typeBuilder.cpp]: More char* leaking

File: lib/ObjWriter/typeBuilder.h [hunk 3]
```diff
@@ -67,14 +75,22 @@ class UserDefinedTypesBuilder {
...
                                 EnumRecordTypeDescriptor *TypeRecords);
+  unsigned GetPointerType(TypeIndex ClassIndex);
+
+  void AddBaseClass(FieldListRecordBuilder &FLBR, unsigned BaseClassId);
 
   MCObjectStreamer *Streamer;
   BumpPtrAllocator Allocator;
   TypeTableBuilder TypeTable;
+
+  unsigned TargetPointerSize;
```
Comment (nattress) [inline lib/ObjWriter/typeBuilder.h]: This should get initialized to 0 in the constructor to be safe.

File: lib/ObjWriter/typeBuilder.cpp [hunk 4]
```diff
@@ -130,10 +132,91 @@ unsigned UserDefinedTypesBuilder::GetCompleteClassTypeIndex(
...
+      ss >> BoundsName;
+      DataMemberRecord BoundsDMR(Access, TypeIndex(SimpleTypeKind::Int32),
+                                 Offset, BoundsName);
+      FLBR.writeMemberType(BoundsDMR);
+      FieldsCount++;
+      Offset += 4;
+    }
+  }
+
+  ArrayRecord AR(ElementTypeIndex, IndexType, ArrayDescriptor.Size, "");
```
Comment (nattress) [inline lib/ObjWriter/typeBuilder.cpp]: Should we set the Name field to something useful here? I'm not very familiar with CodeView - does this feed some friendly array type name in the debugger?

Comment (sandreenko) [inline lib/ObjWriter/typeBuilder.cpp]: No, it is the name of arrayElement + arrayIndex descriptor. ProjectN inserts empty string there too.

File: lib/ObjWriter/typeBuilder.h (hunk 3 above)
Comment (sandreenko) [inline lib/ObjWriter/typeBuilder.h]: Fixed

File: lib/ObjWriter/typeBuilder.cpp (hunk 2 above)
Comment (sandreenko) [inline lib/ObjWriter/typeBuilder.cpp]: Fixed

File: lib/ObjWriter/typeBuilder.cpp (hunk 1 above)
Comment (sandreenko) [inline lib/ObjWriter/typeBuilder.cpp]: Fixed

**Comment by sandreenko** [comment]
@nattress could you please look again?

File: lib/ObjWriter/typeBuilder.cpp [hunk 5]
```diff
@@ -130,10 +135,111 @@ unsigned UserDefinedTypesBuilder::GetCompleteClassTypeIndex(
...
+}
+
+const char *ArrayDimentionsDescriptor::GetBoundsName(unsigned index) {
+  if (Bounds.size() <= index) {
+    Resize(index);
+  }
+  return Bounds[index].c_str();
+}
+
+void ArrayDimentionsDescriptor::Resize(unsigned NewSize) {
```
Comment (nattress) [inline lib/ObjWriter/typeBuilder.cpp]: Typo: `ArrayDimentionsDescriptor` should be `ArrayDimensionsDescriptor`

Comment (sandreenko) [inline lib/ObjWriter/typeBuilder.cpp]: Thanks for your review.

**Pull request merged**